ws-server/src/future.rs

```diff
@@ -130,3 +136,72 @@ where
 		this.selector.poll_unpin(cx)
 	}
 }
+
+#[derive(Debug)]
+struct MonitorInner {
+	shutdown_requested: AtomicBool,
+	waker: AtomicWaker,
+}
+
+/// Monitor for checking whether the server has been flagged to shut down.
+#[derive(Debug, Clone)]
+pub(crate) struct StopMonitor(Arc<MonitorInner>);
+
+impl Drop for StopMonitor {
+	fn drop(&mut self) {
+		if Arc::strong_count(&self.0) == 1 {
+			self.0.waker.wake();
+		}
+	}
+}
+
+impl StopMonitor {
+	pub(crate) fn new() -> Self {
+		StopMonitor(Arc::new(MonitorInner { shutdown_requested: AtomicBool::new(false), waker: AtomicWaker::new() }))
+	}
+
+	pub(crate) fn shutdown_requested(&self) -> bool {
+		self.0.shutdown_requested.load(Ordering::Relaxed)
+	}
+
+	pub(crate) fn handle(&self) -> StopHandle {
+		StopHandle(Arc::downgrade(&self.0))
+	}
+}
+
+/// Handle that is able to stop the running server.
+#[derive(Debug, Clone)]
+pub struct StopHandle(Weak<MonitorInner>);
+
+impl StopHandle {
+	/// Requests server to stop. Returns an error if server was already stopped.
+	///
+	/// Returns a future that can be awaited for when the server shuts down.
+	pub fn stop(self) -> Result<ShutdownWaiter, Error> {
+		if let Some(arc) = Weak::upgrade(&self.0) {
+			// We proceed only if the previous value of the flag was `false`
+			if arc.shutdown_requested.swap(true, Ordering::Relaxed) == false {
+				return Ok(ShutdownWaiter(self.0));
+			}
+		}
+		Err(Error::AlreadyStopped)
+	}
+}
+
+/// A `Future` that resolves once the server has stopped.
+#[derive(Debug)]
+pub struct ShutdownWaiter(Weak<MonitorInner>);
+
+impl Future for ShutdownWaiter {
+	type Output = ();
+
+	fn poll(self: Pin<&mut Self>, cx: &mut Context) -> Poll<Self::Output> {
+		match Weak::upgrade(&self.0) {
+			None => Poll::Ready(()),
+			Some(arc) => {
+				arc.waker.register(cx.waker());
```

Comment:
Similar to the above, is there a chance that all of the `StopMonitor`s are dropped after we upgrade our ref here, and before we let go of it again? This would mean that none of the `StopMonitor`s would wake the waker and we'd be stuck forever I think.

Reply:
One approach might be to not rely on just the Arc, and to keep track of the number of `stop_waiters` separately with an AtomicUsize or something. That way, you don't have to worry about these (probably very unlikely) issues I think, since nothing would be (even for a little bit) bumping that count up again when it shouldn't?

Reply:
(I had a think about whether you could do all of this with a `tokio::Semaphore`, but you'd run into an issue telling things to shutdown (by closing the Semaphore) because that would immediately notify any `ShutdownMonitors` that the thing has closed and wouldn't wait for tidy up)

Reply:
(Maybe a simpler solution: after registering the waker, drop the upgraded `arc` and then check the `strong_count` again; if it's 0 then return Ready immediately?)

Reply:
@jsdw did the last suggestion now :)